### ISSUE TITLE
Not all components work

### ISSUE THREAD

**Issue by ehuss** [kennytm/rustup-toolchain-install-master#15]
Some components do not install because the name of the component is different from the directory name.  For example, the `rls` component is actually `rls-preview`.
```
rustup-toolchain-install-master --component=rls
```
for me will fetch https://s3-us-west-1.amazonaws.com/rust-lang-ci2/rustc-builds/0a4a4ffc69f2d05eb8b8a32eaf9bd0607b69fe38/rls-nightly-x86_64-apple-darwin.tar.xz which has these files:
```
rls-nightly-x86_64-apple-darwin
rls-nightly-x86_64-apple-darwin/LICENSE-APACHE
rls-nightly-x86_64-apple-darwin/LICENSE-MIT
rls-nightly-x86_64-apple-darwin/README.md
rls-nightly-x86_64-apple-darwin/components
rls-nightly-x86_64-apple-darwin/install.sh
rls-nightly-x86_64-apple-darwin/rls-preview
rls-nightly-x86_64-apple-darwin/rls-preview/bin
rls-nightly-x86_64-apple-darwin/rls-preview/bin/rls
rls-nightly-x86_64-apple-darwin/rls-preview/manifest.in
rls-nightly-x86_64-apple-darwin/rls-preview/share
rls-nightly-x86_64-apple-darwin/rls-preview/share/doc
rls-nightly-x86_64-apple-darwin/rls-preview/share/doc/rls
rls-nightly-x86_64-apple-darwin/rls-preview/share/doc/rls/LICENSE-APACHE
rls-nightly-x86_64-apple-darwin/rls-preview/share/doc/rls/LICENSE-MIT
rls-nightly-x86_64-apple-darwin/rls-preview/share/doc/rls/README.md
rls-nightly-x86_64-apple-darwin/rust-installer-version
rls-nightly-x86_64-apple-darwin/version
```

Notice the directory is `rls-preview`, not `rls`.

I would guess that the solution is to first read the `components` file to get the list of directories that exist.  I'm not terribly familiar with how rustup works, but that seems like it would work.

(I can try to help implement that if it seems like a reasonable approach.)

**Comment by lzutao** [comment]
cc @kinnison

**Comment by kinnison** [comment]
> cc @kinnison

Not sure why you're cc'ing me @lzutao

**Comment by lzutao** [comment]
Sorry! I should have to be more clear. 
Do you have any ideas on how to extract these components tarballs just like rustup?

**Comment by kinnison** [comment]
Rustup unpacks the tarballs, ignoring the first path element, otherwise it's a pretty normal unpack.  You can see the code here: https://github.com/rust-lang/rustup.rs/blob/master/src/dist/component/package.rs#L271

**Comment by lzutao** [comment]
I have a fix in #32
